### ISSUE TITLE
WebSocket

### ISSUE THREAD

**Issue by ronag** [nodejs/undici#932]
https://fetch.spec.whatwg.org/#websocket-protocol

**Comment by mcollina** [comment]
cc @lpinca

**Comment by szmarczak** [comment]
This should be quite easy. Reference: https://github.com/szmarczak/http2-wrapper/blob/master/examples/ws/index.js

**Comment by lpinca** [comment]
@szmarczak I know it is a demo but that example basically skips the opening handshake. There is no validation and no subprotocol/extension negotiation.

**Comment by jodevsa** [comment]
@mcollina  this would be really a great opportunity for me to learn more about WS. Can I start working on this?

**Comment by mcollina** [comment]
I don't think we should do another implementation, I think the `ws` module is great, we might just pull it in as a dep.

cc @lpinca @KhafraDev

**Comment by KhafraDev** [comment]
ws is great but I think the implementation is *too different* from the spec WebSocket. There's also outstanding issues where it uses a custom EventTarget (https://github.com/websockets/ws/issues/1818). Then there's a bunch of webidl issues I could spot with a cursory glance (Websocket extends EventEmitter, underscore indexed properties) and the fact that WPTs aren't being run.

Undici (or maybe the nodejs org) should fork or add ws as a git submodule where we could make breaking changes from the ws api & run WPTs, to ensure cross compatibility.

**Comment by lpinca** [comment]
Following the https://websockets.spec.whatwg.org/#websocket-protocol spec in every single detail is a no goal for `ws`. It is limiting and there is no real advantage in my opinion.

I think [`jsdom`](https://github.com/jsdom/jsdom) runs websockets WPT tests. They use `ws` under the hood with some minor adjustments.

**Comment by mcollina** [comment]
I think we should have our own implementation then. Maybe reusing  ws or some parts of it under the hood.

**Comment by jodevsa** [comment]
@mcollina i will give it a try and try to formulate a plan before implementing so that we don't waste time

**Comment by jodevsa** [comment]
> I think we should have our own implementation then. Maybe reusing ws or some parts of it under the hood.
I'm still not quite sure about how we want to implement it. Do we wanna use ws as a dependency inside undici or do we want to implement WS from scratch and maybe get some ideas from ws?

**Comment by mcollina** [comment]
Wdyt @KhafraDev?

**Comment by KhafraDev** [comment]
Using ws:
- uses http(s) under the hood
- no undici Dispatcher
- less security risk

Creating our own:
- faster websockets (ws uses node http(s) under the hood)
- can use undici dispatchers (do websockets work with proxies?)
- more potential security risks

If possible, having our own implementation that borrows from ws would be the ideal solution. We would need to implement the entirety of the WebSocket class in either case, for the reasons mentioned above ("Following the ... spec ... is a non-goal for ws").

**Comment by lpinca** [comment]
> - faster websockets (probably)

The WebSocket protocol has nothing to do with HTTP apart from the initial handshake. Once the initial connection is established the work is done against a `{net,tls}.Socket`. I'm sure a new implementation would be faster but it won't be faster just because you can reuse something in `undici`. You have to write a faster parser and message writer. Also, the parser and message writer are not the bottlenecks in `ws`. The bottleneck is `net.Socket`. Of course, `ws` adds some overhead over a plain `net.Socket`, that is inevitable. A new implementation can try to reduce that overhead, for example by corking writes, but I think it will not be much faster.

> - can use undici dispatchers (do websockets work with proxies?)

Yes, they do, but not all HTTP proxies can proxy WebSocket.

---

`ws` was not designed to be used in `fetch`. This part of the WHATWG spec https://websockets.spec.whatwg.org/#websocket-protocol alters [RFC 6455](https://www.rfc-editor.org/rfc/rfc6455) and refers to browser related things (CSP, HSTS) that do not make sense in a server side context. It should be implemented from scratch.

Other `ws` deviations from the WHATWG spec that I can think of are:

- The `WebSocket` class inherits from `EventEmitter` instead of `EventTarget`. Inheriting from `EventTarget` negatively affects performance and is limiting.
- No direct `Blob` supports. There is no `Blob` support for `Writable.prototype.write()` in Node.js.
- Custom additions and events.

**Comment by KhafraDev** [comment]
I took a shot at it and got close (I think) to a working initial handshake. If anyone wants a place to start, it might be worth checking my branch out https://github.com/KhafraDev/undici/tree/ws/lib/websocket. It doesn't currently work btw (the `onUpgrade` callback is never called).

**Comment by jimmywarting** [comment]
👍  for rolling our own that is more spec compatible with wpt tests

**Comment by KhafraDev** [comment]
I took another stab at the initial handshake and got it working! The mistake was really dumb on my part (`sha1` is a supported hash, not `SHA-1` and some error swallowing).

```js
import { obtainWebSocketConnection } from './lib/websocket/connection.js'
import { getGlobalDispatcher } from './index.js'

const socket = await obtainWebSocketConnection(
  new URL('wss://echo.websocket.events/'),
  getGlobalDispatcher()
)

console.log('connected', !!socket)
```

Anyone is free to work/contribute to my branch - help would be appreciated(!!). I'm making some progress, but there is still *a **lot*** left to do.

**Comment by jodevsa** [comment]
Hi @KhafraDev,

great start! I'm really interested in working with you on this. I hope I can make some progress today; I'll try to leave some handovers in the PR comments

**Comment by jodevsa** [comment]
Hi @KhafraDev,


I was able to implement sending TEXT data and do the masking logic; we are currently able to send and receive text messages through WS: 
<img width="776" alt="Screenshot 2022-12-04 at 01 16 21" src="https://user-images.githubusercontent.com/14154314/205467722-3663e31b-8869-412e-828b-2acb637de0cd.png">


please have a look at my PR: https://github.com/KhafraDev/undici/pull/30
